src/mds/LogSegment.h

```diff
@@ -50,6 +51,13 @@ class LogSegment {
     dirty_dirfrag_nest(member_offset(CInode, item_dirty_dirfrag_nest)),
     dirty_dirfrag_dirfragtree(member_offset(CInode, item_dirty_dirfrag_dirfragtree))
   {}
+ public:
+  using seq_t = uint64_t;
+
+  [[nodiscard]] static AutoSharedLogSegment create(uint64_t _seq, loff_t off = -1)
+  {
+    return std::shared_ptr<LogSegment>(new LogSegment(_seq, off));
```

Comment:
https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared

Reply:
`make_shared` is inaccessible here because the constructor is private. See class `Best()` in the example from the [documentation on shared_ptr](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)

Additionally, `make_shared` comes at the cost of pinning the memory of the whole object when only weak pointers are left since the same memory block is used to hold the shared pointer control and the class data. This is mentioned in the second note on the documentation page you've referenced.

Reply:
> `make_shared` is inaccessible here because the constructor is private. See class `Best()` in the example from the [documentation on shared_ptr](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)

ok

> Additionally, `make_shared` comes at the cost of pinning the memory of the whole object when only weak pointers are left since the same memory block is used to hold the shared pointer control and the class data. This is mentioned in the second note on the documentation page you've referenced. To clarify, I'm not sure this will be a blocker for us, and maybe we could even benefit from it, but I think it's an important point to keep in mind

I don't really think that will be significant but I also wonder if a clever implementation is permitted to `realloc` the memory block when `T` is destroyed. What an odd caveat I had the opposite understanding about.